include/nbl/builtin/hlsl/emulated/emulated_float64_t_impl.hlsl

```diff
@@ -0,0 +1,327 @@
+
+// TODO: when it will be possible, use this unions wherever they fit:
+/*
+* union Mantissa
+* {
+*   struct
+*   {
+*       uint32_t highBits;
+*       uint64_t lowBits;
+*   };
+*
+*   uint32_t2 packed;
+* };
+*
+*/
+/*
+* union Mantissa
+* {
+*   struct
+*   {
+*       uint64_t lhs;
+*       uint64_t rhs;
+*   };
+*
+*   uint32_t4 packed;
+* };
+*
+*/
+
+namespace nbl
+{
+namespace hlsl
+{
+namespace impl
+{
+NBL_CONSTEXPR_INLINE_FUNC uint64_t2 shiftMantissaLeftBy53(uint64_t mantissa64)
+{
+    uint64_t2 output;
+    output.x = mantissa64 >> (64 - ieee754::traits<float64_t>::mantissaBitCnt);
+    output.y = mantissa64 << (ieee754::traits<float64_t>::mantissaBitCnt);
+
+    return output;
+}
+
+NBL_CONSTEXPR_INLINE_FUNC uint64_t packFloat64(uint32_t zSign, int zExp, uint32_t zFrac0, uint32_t zFrac1)
+{
+    uint32_t2 z;
+
+    z.x = zSign + (uint32_t(zExp) << 20) + zFrac0;
+    z.y = zFrac1;
+
+    uint64_t output = 0u;
+    output |= (uint64_t(z.x) << 32) & 0xFFFFFFFF00000000ull;
+    output |= uint64_t(z.y);
+    return  output;
+}
+
+NBL_CONSTEXPR_INLINE_FUNC uint32_t2 packUint64(uint64_t val)
+{
+    return uint32_t2((val & 0xFFFFFFFF00000000ull) >> 32, val & 0x00000000FFFFFFFFull);
+}
+
+NBL_CONSTEXPR_INLINE_FUNC uint64_t unpackUint64(uint32_t2 val)
+{
+    return ((uint64_t(val.x) & 0x00000000FFFFFFFFull) << 32) | uint64_t(val.y);
+}
+
+template<bool FlushDenormToZero>
+inline uint64_t castFloat32ToStorageType(float32_t val)
+{
+    if (FlushDenormToZero)
+    {
+        const uint64_t sign = uint64_t(ieee754::extractSign(val)) << 63;
+        if (tgmath::isinf(val))
+            return ieee754::traits<float64_t>::inf | sign;
+        uint32_t asUint = ieee754::impl::bitCastToUintType(val);
+        const int f32BiasedExp = ieee754::extractBiasedExponent(val);
+        if (f32BiasedExp == 0)
+            return sign;
+        const uint64_t biasedExp = uint64_t(f32BiasedExp - ieee754::traits<float32_t>::exponentBias + ieee754::traits<float64_t>::exponentBias) << (ieee754::traits<float64_t>::mantissaBitCnt);
+        const uint64_t mantissa = (uint64_t(ieee754::traits<float32_t>::mantissaMask) & asUint) << (ieee754::traits<float64_t>::mantissaBitCnt - ieee754::traits<float32_t>::mantissaBitCnt);
+
+        return sign | biasedExp | mantissa;
+    }
+    else
+    {
+        // static_assert(false);
+        return 0xdeadbeefbadcaffeull;
+    }
+};
+
+NBL_CONSTEXPR_INLINE_FUNC bool isZero(uint64_t val)
+{
+    return (val << 1) == 0;
+}
+
+inline uint64_t castToUint64WithFloat64BitPattern(uint64_t val)
+{
+    if (isZero(val))
+        return val;
+
+#ifndef __HLSL_VERSION
+    int exp = findMSB(val);
+#else
+    int exp = 63;
+    uint64_t mask = ieee754::traits<float64_t>::signMask;
+    while (!(val & mask))
+    {
+        --exp;
+        mask >>= 1;
+    }
+
+
+    //uint32_t2 valPacked = packUint64(val);
+    //int exp = valPacked.x ? firstbithigh(valPacked.x) + 32 : firstbithigh(valPacked.y);
+    //exp = 63 - exp;
```

Comment:
1. Start a `spirv_intrinsics/glsl.std.450.hlsl` header
2. Add the GLSLstd450FindUMsb intrinsic using syntax outlined here https://github.com/microsoft/DirectXShaderCompiler/issues/6751
3. add a `findMSB` and `findLSB` to `glsl_compat` header and handle the HLSL/C++ divergence there


> This instruction is currently limited to 32-bit width components.

You need to do some gymnastics to implement 64-bit variants of findLSB and findMSB in glsl_compat